### ISSUE TITLE
LocalCUDACluster

### ISSUE THREAD

**Issue by mrocklin** [dask/distributed#2430]
Today people using Dask with multiple GPUs on a single machine tend to have a script that does something like the following:

1.  Start a scheduler
2.  For each GPU on the machine
    - Start a new process that ...
    - Set an environment variable like `CUDA_VISIBLE_DEVICES` to restrict that process to one GPU
    - Start a Dask worker within that process

These scripts tend to be error prone and messy.  It would be good to have a single class that did this, similar to `LocalCluster`.  Hopefully also in building this class we can clean up `LocalCluster` a bit (a long-held goal of mine).

I think I can get everything here to work, some questions/changes that might need to be addressed:

-  How do I test that a library like PyTorch or Numba has appropriately read `CUDA_VISIBLE_DEVICES`?
-  How do I determine the number of GPUs available?
-  I think I'll need to add an `env=` keyword to `Nanny`, which I think is a fine leak of GPU specific code into the dask/distributed codebase.
-  I'll probably create a subclass of LocalCluster that overwrites the `_start_worker` method to add the appropriate environment variable to Nanny, and also changes around the defaults and some resource settings.

cc @kkraus14 @sklam @seibert

**Comment by kkraus14** [comment]
> How do I test that a library like PyTorch or Numba has appropriately read CUDA_VISIBLE_DEVICES?

This actually happens at the CUDA level as part of the CUDA API so it's transparent to libraries. That being said, for cuDF and RAPIDS we're working to be able to specify this outside of relying on the environment variables as there's other implications related to things like communication libraries.

**Comment by kkraus14** [comment]
> How do I determine the number of GPUs available?
The easiest way is probably to call `nvidia-smi -L` from the CLI which gives you a list of GPUs along with a UUID for each GPU. I believe this is how Numba detects and enumerates GPUs but Siu or Stan could answer that better than me.

**Comment by mrocklin** [comment]
OK, so I probably can't test this except by running something like a performance test?

> The easiest way is probably to call nvidia-smi -L from the CLI

I take it that there is no standard Python library to query information like this?

**Comment by kkraus14** [comment]
> I take it that there is no standard Python library to query information like this?

Unfortunately there is not at this time.

**Comment by mrocklin** [comment]
Is there an obvious set of functionality that would make sense to put into a small `psutil` style package?  Is there some obvious analog on the C/CUDA side that someone could wrap in a straightforward manner?

I remember sitting down with @lesteve at SciPy last year and he had some Python friendly library that was providing GPU usage statistics.

**Comment by lesteve** [comment]
> I remember sitting down with @lesteve at SciPy last year and he had some Python friendly library that was providing GPU usage statistics.

I think I was actually calling `nvidia-smi` with the right arguments, maybe that was with somebody else? See what I added to the distributed documentation about registering custom metrics:
http://distributed.dask.org/en/latest/web.html#per-worker-resources

**Comment by lesteve** [comment]
Thanks for pinging me, I'll definitely try to follow this, although I am not too sure how that can be useful to people around me at the moment if I am being fully honest.

Let me explain a little bit why: most our GPUs are inside a cluster. The scheduler (SGE in our case) is setting the CUDA_VISIBLE_DEVICES correctly. Also they don't really do that much multi-GPU AFAIK. Basically the main use case ATM is to use dask to do embarassingly parallel computations that are dispatched to multiple GPUs. Generally GPUs may not be on the same machine (if you want multiple GPUs on the same machine you may actually wait longer in the queue to get them). When they did attempt to use multi-GPU, my understanding is that they used high-level PyTorch APIs (`torch.distributed` or `torch.nn.DataParallel` not 100% sure about this).

**Comment by sklam** [comment]
Replying to "How do I determine the number of GPUs available?" chain.

Numba initializes the CUDA driver via ctypes and uses the [driver API](https://docs.nvidia.com/cuda/cuda-driver-api/index.html) to find out the number of GPUs.  A lot of try-except is needed to:
* Make sure there's a real and recent cuda driver (https://github.com/numba/numba/blob/16dc526b2ba99478b4cd161713eade29bff6a3c4/numba/cuda/cudadrv/driver.py#L86).
* Make sure the cuda driver and be init (i.e. `cuInit()`) (https://github.com/numba/numba/blob/16dc526b2ba99478b4cd161713eade29bff6a3c4/numba/cuda/cudadrv/driver.py#L227).
* Make sure you are not re-initializing cuda driver after forking. (https://github.com/numba/numba/blob/16dc526b2ba99478b4cd161713eade29bff6a3c4/numba/cuda/cudadrv/driver.py#L320)

Then, getting the device count is just calling `cuDeviceGetCount`.

I really wish there is a common library that everyone can use to do all these.

**Comment by mrocklin** [comment]
This is now in https://github.com/rapids/dask-cuda .  Not yet on PyPI.  Closing.
